Refactor

f1erro opened the pull request:
One of the big problems with this package is it's unit testability.  There were a few infinite loops that kicked off goroutines which are impossible to kill, which made testing difficult because those goroutines would operate on shared global space.  Also, it creates confusing logs (for example, a network resiliency test will continue to retry for the duration of the test)

This pull request does three things

1) Refactor the entire package to be more readable and understandeable, and more broken up/logically grouped

2) Extract core functionality into unit-testable functions.  For the code driving polling loops, there is also now a kill switch which can be used in the future when we implement a remote command channel.  The immediate use case, however, is for having more control during testing

3) Added support for sqlite snapshots.  To isolate these changes, just check out the first commit in this PR. It's a tiny change, see the function `processSnapshotServerFileResponse`.  It just streams the response to disk and sets the DB version, which when the event is emitted, will be used by the plugins to get a handle on the DB, which will already be there (the full path is created in apid-core, but if the file already exists it will just work)

Tests are coming later.  Right now all of the existing tests pass, but this refactor enables more unit testing to be done, which this package badly needs (particularly the logic across `snapshot.go`, `changes.go`, and `apigee_sync.go`, which was previously all contained in `apigee_sync.go`)

sramamoorthy73 commented on init.go:
```diff
@@ -156,7 +156,11 @@ func postInitPlugins(event apid.Event) {
 
 		tokenManager = createTokenManager()
 
-		go bootstrap()
```
Where is the new bootstrap() function definition. Are all the changes in the code review?

f1erro replied:
bootstrap function is still in apigee_sync. The diff should be there.  All of the changes are in code review.

theganyo replied:
I don't see it in the diffs here either.

sramamoorthy73 replied:
Yeah. It was too big a change that It did not expose the changes :)

f1erro replied:
ah, yes. lots of deletion ;)

apg-pk commented on changes.go:
```diff
@@ -0,0 +1,207 @@
+
+/*
+ * Long polls the change agent with a 45 second block. Parses the response from
+ * change agent and raises an event. Called by pollWithRetry().
```
Misleading comment. Did you mean pollWithBackoff()?

f1erro replied:
Yes, good catch.  I need to go back and comb through small things like that

theganyo commented on backoff.go:
```diff
@@ -0,0 +1,82 @@
+package apidApigeeSync
```
Add tests for backoff.go.

theganyo commented:
I assume the test in apigeeSync_suite_test.go will be able to be moved into a real test as well.

haomingzhang commented on snapshot.go:
```diff
@@ -0,0 +1,254 @@
+
+	done := make(chan bool)
+	log.Info("Emitting Snapshot to plugins")
+	events.EmitWithCallback(ApigeeSyncEventSelector, snapshot, func(event apid.Event) {
```
Try to pull the new apid-core, and use the new blocking Emit() with returned channel for this:

```
select {
 	case <-time.After(pluginTimeout):
 		log.Panic("Timeout. Plugins failed to respond to snapshot.")
 	case <-events.Emit(ApigeeSyncEventSelector, snapshot):
}
```
No need for the `done` chan

f1erro replied:
nice, thanks!

haomingzhang commented on changes.go:
```diff
@@ -0,0 +1,207 @@
+			done <- true
+		})
+
+		select {
```
Try to pull the new apid-core, and use the new blocking Emit() with returned channel for this:
```
select {
 	case <-time.After(httpTimeout):
 		log.Panic("Timeout. Plugins failed to respond to changes.")
 	case <-events.Emit(ApigeeSyncEventSelector, &resp):
}
```
No need for the `done` chan

f1erro replied:
done

haomingzhang commented on changes.go:
```diff
@@ -0,0 +1,207 @@
+
+func changesRequireDDLSync(changes common.ChangeList) bool {
+
+	return !mapIsSubset(knownTables, extractTablesFromChangelist(changes))
```
It's not necessary to make a map from `changes.Changes`. You already have `knownTables` as a map, so you can just iterate over the` ChangeList` and try to hit each change in `knownTables`. No need for making the 2nd map.

f1erro replied:
good catch. tunnel vision

haomingzhang commented on changes.go:
```diff
@@ -0,0 +1,207 @@
+			}
+		}
+
+		r.Body.Close()
```
I would suggest move `r.Body.Close()` to immediately following where you created `r`, and use `defer r.Body.Close()`.
This is because all the `return` above did not close the body before return.

f1erro replied:
done

haomingzhang commented on changes.go:
```diff
@@ -0,0 +1,207 @@
+	lastSequence = getLastSequence()
+
+	for {
+		select {
```
The behavior of this select is:
If `quit` is not available, it goes into `default`. Since `getChanges ` is blocking (using http), when receiving `quit`, this func will not return immediately if it is blocked in `getChanges`. It will not quit until it finishes `getChanges `, in the worst case, after the timeout `httpTimeout + httpTimeout`.
I'm not sure what is your expected behavior, but if you want the `quit` to take effect immediately, I would call `getChanges` in another go routine like this:
```
done := make(chan error, 1)
go getChanges(changesUri, done)
for {
  select {
  case <- quit:
    return quitSignalError{}
  case err := <- done:
    return err
  }
}

```

f1erro replied:
This is a good point, but in this suggested code, if the quit signal is received, the `getChanges` goroutine will still be running.  The purpose of this quit is to ensure that no goroutines are left lying around after `pollWithBackoff` returns so that you can be sure there are no goroutines still running which could modify global state, and thus impair testing.  Also, when we implement a command/control channel, this might be a guarantee we need before sending a confirmation back.  For instance, you send a "stop polling" command.  By the time that we return a response, we should guarantee there aren't any routines/processes running in that capacity

haomingzhang replied:
OK, this is reasonable.

f1erro commented:
new commit includes
```- bug fix for token manager not returning correct `expected200Error`
- added channel quit signals to the sync test so that after the tests close, output doesn't get intermingled with the remaining tests
- moving initial test out of `BeforeSuite` into a regular test
- slight refactor of `init.go`
- Implemented more clean slate logic to ensure each test does not conflict```

f1erro commented:
One token test is timing out, has never happened on my machine, going to investigate why

sramamoorthy73 commented on token_test.go:
```diff
@@ -65,7 +65,7 @@ var _ = Describe("token", func() {
 
 		It("should get a valid token", func() {
-
+log.Info("\n\n\n\nHERE\n\n")
```
Mitch, This log was perhaps used for debugging. Is this MR fixing the race issue?

theganyo commented:
It looks like we're at 68.7% overall statement coverage now. This is probably because we now set `ReliableAPI:  true` for the mock server - which will obviously avoid a lot of the error conditions. We should to test error conditions explicitly, but if not we should think about allowing the mock server to generate errors again. In addition, we should create a backoff_test.go to test the backoff module explicitly.

theganyo commented on apigee_sync_test.go:
```diff
@@ -5,90 +5,283 @@ import (
-		expectedTables["kms.company"] = true
-		expectedTables["edgex.apid_cluster"] = true
-		expectedTables["edgex.data_scope"] = true
+		var initializeContext = func() {
```
Is it necessary to make these explicit functions instead of using beforeEach() and afterEach()?

f1erro replied:
I think so, because we don't need to spin up the mock server for every test, only those that need it, which are the sync tests.

theganyo replied:
They could be separated into separate contexts...

f1erro replied:
If your beforeEach method is declared outside a context, as it currently is, you can only have one.  If you put a beforeEach in each context, then every other context has to have a beforeEach.  Not sure which is actually better, but personally I prefer consolidating the beforeEach, and doing the sync specific stuff locally/manually

theganyo commented on apigee_sync_test.go:
```diff
@@ -5,90 +5,283 @@ import (
+			var closeDone <-chan bool
+			initializeContext()
+			// do not wipe DB after.  Lets use it
+			wipeDBAferTest = false
```
This looks like a side-effect... Does it assume tests run in a given order?

f1erro replied:
yes.  Top level containers run in a random order, but within a container, they run in a given order.  So with the sync tests, we run a clean slate bootstrap first, then run a test starting from an existing snapshot, and use this flag to coordinate that

theganyo replied:
Ok. It feels a bit icky to me.

f1erro replied:
yeah.  the alternative is to have a very large messy test that does a full clean slate start (thus duplicating the previous test), a shutdown, then another start

I think that if you consider this pair of test cases a special tightly bound test case, maybe it's less icky.

But I agree.
